### PR TITLE
Destinations CDK: refreshes logic

### PR DESCRIPTION
**Pull request by edgao**
## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by edgao** [comment]
* **#38616** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/38616?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#38622** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/38622?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @edgao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/StorageOperation.kt [hunk 1]
```diff
@@ -16,9 +15,38 @@ interface StorageOperation<Data> {
...
-     * Prepare staging area which cloud be creating any object storage, temp tables or file storage
+     * Prepare staging area which cloud be creating any object storage, temp tables or file storage.
+     * Similar to [createFinalTable], accepts a [suffix] parameter, which should be used in conjunction
+     * with [overwriteStage].
+     *
+     * @param replace If true, then replace existing resources with empty e.g. tables. If false,
+     * then leave existing resources untouched.
      */
-    fun prepareStage(streamId: StreamId, destinationSyncMode: DestinationSyncMode)
+    fun prepareStage(streamId: StreamId, suffix: String, replace: Boolean = false)
```
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/StorageOperation.kt:25]: hm. just realized that this is weirdly inconsistent

* prepareStage (and friends) all take a `suffix`
* but `writeToStage` expects the caller to modify the StreamId (i.e. `streamId.copy(rawName = streamId.rawName + suffix)`)

after doing the bigquery implementation, I actually kind of prefer passing an explicit suffix, but not strongly - @airbytehq/destinations wdyt?

**Review by gisripa — APPROVED** [review]

File: airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt [hunk 2]
```diff
@@ -47,9 +68,103 @@ abstract class AbstractStreamOperation<DestinationState : MinimumDestinationStat
...
+            // In this case, we MUST retain the records from the truncate attempt.)
+            storageOperation.transferFromTempStage(stream.id, TMP_TABLE_SUFFIX)
+
+            // We need to combine the raw table statuses from the real and temp raw tables.
+            val hasUnprocessedRecords =
+                destinationInitialStatus.initialTempRawTableStatus.hasUnprocessedRecords ||
+                    destinationInitialStatus.initialRawTableStatus.hasUnprocessedRecords
+            // Pick the earlier min timestamp.
+            val maxProcessedTimestamp: Optional<Instant> =
+                destinationInitialStatus.initialRawTableStatus.maxProcessedTimestamp
```
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:99]: this seems very complex to just get the min timestamp (if I'm understanding this correcttly)

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:99]: yeah, I'd love a suggestion on how to do this better :/ 

(also, getting things prepped for monday release - any concern with merging this as-is + fixing in a followup?)

Comment (edgao) [reply airbyte-cdk/java/airbyte-cdk/typing-deduping/src/main/kotlin/io/airbyte/integrations/base/destination/operation/AbstractStreamOperation.kt:99]: going to just leave this as-is for now. there's probably something along these lines, though we probably don't _actually_ want to introduce [liftA2](https://hackage.haskell.org/package/base-4.20.0.1/docs/Prelude.html#v:liftA2) to kotlin :P

image

**Pull request merged**